protocols/tls_cryptolib/src/lib.rs

```diff
@@ -0,0 +1,529 @@
+}
+
+// === Hashing === //
+
+pub fn hash(ha: &HashAlgorithm, payload: &ByteSeq) -> Result<Digest, CryptoError> {
+    match ha {
+        HashAlgorithm::SHA256 => Result::<Digest, CryptoError>::Ok(Digest::from_seq(&sha256(payload))),
+        HashAlgorithm::SHA384 => Result::<Digest, CryptoError>::Err(CryptoError::UnsupportedAlgorithm),
+    }
+}
+
+// === HMAC === //
+
+pub fn hmac_tag(ha: &HashAlgorithm, mk: &MacKey, payload: &ByteSeq) -> Result<HMAC, CryptoError> {
+    match ha {
+        HashAlgorithm::SHA256 => Result::<HMAC, CryptoError>::Ok(HMAC::from_seq(&hmac(mk, payload))),
+        HashAlgorithm::SHA384 => Result::<HMAC, CryptoError>::Err(CryptoError::UnsupportedAlgorithm),
+    }
+}
+
+fn check_tag_len(a: &HMAC, b: &HMAC) -> Result<(), CryptoError> {
+    if a.len() == b.len() {
+        Result::<(), CryptoError>::Ok(())
+    } else {
+        Result::<(), CryptoError>::Err(CryptoError::MacFailed)
+    }
+}
+
+fn check_bytes(a: U8, b: U8) -> Result<(), CryptoError> {
+    if !a.equal(b) {
+        Result::<(), CryptoError>::Err(CryptoError::MacFailed)
+    } else {
+        Result::<(), CryptoError>::Ok(())
+    }
+}
+
+pub fn hmac_verify(
+    ha: &HashAlgorithm,
+    mk: &MacKey,
+    payload: &ByteSeq,
+    m: &HMAC,
+) -> Result<(), CryptoError> {
+    let my_hmac = hmac_tag(ha, mk, payload)?;
+    check_tag_len(m, &my_hmac)?;
+    for i in 0..m.len() {
+        check_bytes(my_hmac[i], m[i])?;
+    }
+    Result::<(), CryptoError>::Ok(())
+}
+
+// === Signatures === //
+
+bytes!(EcOidTag, 9);
+
+// Some ASN.1 helper functions
+fn get_length_length(b: &ByteSeq) -> usize {
+    if U8::declassify(b[0]) >> 7 == 1u8 {
+        declassify_usize_from_U8(b[0] & U8(0x7fu8))
+    } else {
+        0
+    }
+}
+fn get_length(b: &ByteSeq, len: usize) -> usize {
+    declassify_u32_from_U32(U32_from_be_bytes(U32Word::from_slice(b, 0, len))) as usize
+        >> ((4 - len) * 8)
+}
+fn get_short_length(b: &ByteSeq) -> usize {
+    declassify_usize_from_U8(b[0] & U8(0x7fu8))
+}
+
+// Very basic ASN.1 parser to read the ECDSA public key from an X.509 certificate.
+pub fn verification_key_from_cert(cert: &ByteSeq) -> Result<VerificationKey, CryptoError> {
+    // cert is an ASN.1 sequence. Take the first sequence inside the outer.
+    // Skip 1 + length bytes
+    let skip = 2 + get_length_length(&cert.slice_range(1..cert.len())) + 1;
+    let seq1_len_len = get_length_length(&cert.slice_range(skip..cert.len()));
+    let skip = skip + 1;
+    let seq1_len = get_length(&cert.slice(skip, cert.len() - skip), seq1_len_len);
+    let mut seq1 = cert.slice_range(skip + seq1_len_len..skip + seq1_len_len + seq1_len);
+
+    // Read sequences until we find the ecPublicKey (we don't support anything else right now)
+    let mut pk = VerificationKey::new(0);
+    // XXX: the typechecker should allow _ here. We don't need the counter.
+    for _i in 0..seq1.len() {
+        // FIXME: we really need a break statement.
```

Comment:
Unfortunately `break` statements are really hostile to formal verification... The next least difficult thing to add to improve loops in hacspec would be a while loop, but then you need to declare loop invariants.

Reply:
Yeah, I think we need to chat a little about this. We either need a better way to parse byte sequences with for loops, or we need better loops. This is a while loop forced into a for loop without breaking, which obviously doesn't look good. I started to add [issues to the book](https://github.com/hacspec/book/issues/) to describe there how to idiomatically do stuff like this. Right now it seems there is no *good* way to do it though.

Reply:
Yes I've seen https://github.com/hacspec/book/issues/3 :) Do you think allowing while loops would provide a good way for these situations? @karthikbhargavan what's your opinion on adding while loops in hacspec?

Reply:
A loop like this one is essentially `while X not found` and then walk through the bytes. While loops is the obvious choice here. But another option would be to have better chunking in `Seq` that allows parsing of tag-len encoded byte sequences.